src/Type.hs

```diff
@@ -214,30 +212,23 @@ checkContract store constructors (U.Contract constr@(U.Definition _ cid _ _ _ _
 -- checks a transition given a typing of its storage variables
 checkTransition :: Env -> U.Transition -> Err [Behaviour]
 checkTransition env (U.Transition _ name contract iface@(Interface _ decls) iffs cases posts) =
-  noIllegalWilds *>
   -- constrain integer calldata variables (TODO: other types)
   fmap fmap (makeBehv <$> checkIffs env' iffs <*> traverse (checkExpr env' SBoolean) posts)
-  <*> traverse (checkCase env') normalizedCases
+  <*> traverse (checkCase env') elaborateCases
   where
     env' = addCalldata env decls
 
-    noIllegalWilds :: Err ()
-    noIllegalWilds = case cases of
-      U.Direct   _  -> pure ()
-      U.Branches bs -> for_ (init bs) $ \c@(U.Case p _ _) ->
-                          when (isWild c) (throw (p, "Wildcard pattern must be last case"))  -- TODO test when wildcard isn't last
-
-    -- translate wildcards into negation of other branches and translate a single case to a wildcard
-    normalizedCases :: [U.Case]
-    normalizedCases = case cases of
-      U.Direct   post -> [U.Case nowhere (U.WildExp nowhere) post]
-      U.Branches bs ->
-        let
-          Just (rest, lastCase@(U.Case pn _ post)) = unsnoc bs
-          negation = U.ENot nowhere $
-                        foldl (\acc (U.Case _ e _) -> U.EOr nowhere e acc) (U.BoolLit nowhere False) rest
-        in rest `snoc` (if isWild lastCase then U.Case pn negation post else lastCase)
-    -- TODO ensure non-overlapping and exhaustiveness (maybe with elaboration and mandatory wildcard?) 
+    -- make cases non overlapping by negating previous cases that haven't matched
+    elaborateCases :: [U.Case]
```

Comment:
I was puzzled over this when I was writing this PR. Wildcards are not part of the abstract syntax right now, because the condition in wildcard cases is represented using the literal `True`. 

But the change to the parser in this PR only allows wildcards to be last, so the check is not needed anymore, as it is impossible to have a wildcard in the middle of case analysis.